Switch to Containerless Setup

bwintermann opened the pull request:
Following #25, we start adding support for a non-docker based setup. We utilize Poetry for package management, `click` for command line parsing and `rich` for formatting and better tracebacks. 

**The old Docker/Singularity setup will be removed completely in favor of this.**

**Multiple steps are required to complete this setup:**
- [x] Move all requirements and Python-dependencies to `pyproject.toml`
- [x] Move flake8, black, pytest, etc. configurations to `pyproject.toml`
- [x] Add a executable / runnable script as the new entrypoint (`run_finn.py`, marked as a runnable script in `pyproject.toml`)
- [x] Finish the dependency subcommand 
- [x] Finish the `run` and `build` subcommands
- [x] Replace hardcoded dependency locations in FINN source with a function that can find the dependencies automatically
- [x] Verilator automatic installation
- [x] Replace all `shell=True` args with `shell=False` and cleaned input

**Validation:**
- [x] Full test suite passing

**Optional:**
- [x] Add a user guide to the README, and ideally include some quick guide markdown as a subcommand
- [x] Remove files that are no longer needed (`entrypoint.sh`, `(quick)test.sh`, `setup.cfg`, `setup.py`, etc.)
- [ ] Automatic CI builds that produce Python wheels to easily install FINN?
- [ ] `Inspect` command?

bwintermann commented:
The core functionality is implemented now. I think the best way to proceed is to keep this PR open and add new features when necessary but until then we only merge the bare minimum. 

@iksnagreb I have added the attention-hlslib to the dependencies. Could you please take a look in `interface/manage_deps.py` and `interface/manage_envvars.py` and `pyproject.toml` if you need anything else in terms of dependencies or environment variables? (I also already exchanged the paths in `hlsbackend.py` but don't know if there are any other changes you need).

@LinusJungemann Since you are already working on this, could you check what setup flake8, black etc. need to move from `setup.cfg` to `pyproject.toml`? I have already moved the `pytest` configuration.

LinusJungemann commented:
Just added the isort config to the pyproject.toml. One problem that we will have to deal with is flake8. flake8 does not support pyproject.toml, therefore we will either have to add a thirdparty lib that adds support like flakehell, Flake8-pyproject, flake9, flake518 or move to ruff...

iksnagreb commented:
> @iksnagreb I have added the attention-hlslib to the dependencies. Could you please take a look in `interface/manage_deps.py` and `interface/manage_envvars.py` and `pyproject.toml` if you need anything else in terms of dependencies or environment variables? (I also already exchanged the paths in `hlsbackend.py` but don't know if there are any other changes you need).

Looks good so far, still need to test this with some complete build/setup, though. But this library is almost self-contained, except for some dependencies to the finn-hlslib, and there is nothing which needs to be compiled or generated in advance, so it can be "installed" everywhere as long as the compiler is pointed to search for includes there. SO I do not expect this to break anything.

bwintermann commented:
@LinusJungemann  As much as I like ruff, I think sticking to flake8 is probably best. I'd opt for keeping a `.flake8` config file. At some point we can then deprecate the `setup.cfg`. One more file is probably better than adding _yet another_ dependency for a plugin or replacement. Also this way we can avoid changing the pre-commit config.

@iksnagreb Sounds good!

If we can get the flake8 config ported today, I might be able to run tests and maybe we can merge today or beginning of next week already.

LinusJungemann commented:
I have added the flake config. Currently, the linter fails in several places, but these are all things that are fixed in #44, so I won't fix them here again.

bwintermann commented:
Added some more changes. Whether Vivado/Vitis/HLS tools are at the given paths is now being checked. You can now configure your fallback paths (if you for example have limited space in your home directory, you can place the _default_ dependency location somewhere else). Generally almost everything is now configurable, except for the location of the one `settings.yaml` file which is still in `~/.finn/settings.yaml`. I plan to introduce a list of paths to check though, in case you cannot write to your own home directory.
These configuration options are also writeable via a subcommand `finn config`, with options being

`finn config list`
`finn config get <key>`
`finn config set <key> <value>`

These do what you would guess. 
The tool also warns you, if no environment variable file and settings file exist. For the settings file it offers to write a template one for you if none is found - for the environment variable file not, since a user might have multiple ones and managing these via a text editor is easier than via the `finn` executable. 

All of this we can of course completely rework if it turns out that it doesn't work, this is just a first sketch.

bwintermann commented:
Implemented most of the rework we talked about last week. No more environment variable configuration, settings is automatically found and generally for all settings the order of priorities is:
Comand Line Argument > Environment Variable > Settings File > Defaults in the Code.

I also removed, as discussed, all files related to the old build. Documentation configuration and tutorials are still up. 

I will likely adapt the README tomorrow.

LinusJungemann commented on README.md:
```diff
@@ -29,6 +31,27 @@ The framework is fully open-source in order to give a higher degree of flexibili
+- Python >= 3.10 (<= 3.12)
+- Poetry
+- Vivado, Vitis, Vitis-HLS
+- verilator (will be installed automatically)
```
The verilator line should be removed.
